### ISSUE TITLE
VIDEO - Stuck On "A config file was not found. Would you like to create one now?"

### ISSUE THREAD

**Issue by parkershamblin** [cheat/cheat#571]
Unable to successfully run example "cheat tar" after installing on Windows 10.

**Problem: When I run "cheat tar" it keeps returning "A config file was not found. Would you like to create one now?"**

Here is a [video](https://imgur.com/a/4xRyKO0) I made where I show the exact steps I took while trying to install cheat on my Windows 10 machine.

Any help?

**Comment by chrisallenlane** [comment]
Hi, @parkershamblin 

Thanks for the bug report.

I haven't meaningfully used Windows since the XP days, so I can't help personally, unfortunately. (I still need to assemble a Win10 machine for testing.) That being the case, I'll pin this ticket for visibility. Hopefully someone from the community will be able to help you out.

**Comment by parkershamblin** [comment]
Hi, @chrisallenlane 

I appreciate the help. :)

Looking forward to using your project once someone with a windows machine finds this issue.

**Comment by Benjamin-Connelly** [comment]
>
>
> Looking forward to using your project once someone with a windows machine finds this issue.

Hi @parkershamblin, @chrisallenlane,

I am the least likely person to submit a proper PR for a Windows application, and this is a workaround for `git-bash` shell, not Powershell or CMD.exe. 

The company I work for just released production firmware for the Gateway/Routers we manufacture that replace MS Domain Controllers / Active Directory with all Open Source Software (Samba4, BIND, etc...) on an OpenWRT platform. This said, I was handed my first Windows laptop since Windows 98 to support our customers when they have AD issues. 

I installed the Scoop package manager and during the install 7zip and Git are the first two packages that you install.  I noticed that  `git-bash.exe`  comes with `2.27.0.windows.1` - <https://github.com/ScoopInstaller/Main/blob/master/bucket/git.json> and in my attempt to find a comfortable workflow in a bash shell on Windows, I too saw the same error as you this morning. Here is how I quickly remediated the issue and am using `cheat` in my `git-bash.exe` shell. 

```shell
cp cheat-windows-amd64.exe /usr/bin/
echo alias cheat='/usr/bin/cheat-windows-amd64.exe' >> $HOME/.bashrc
source $HOME/.bashrc
mkdir -p $HOME/.config/cheat && cheat --init > $HOME/.config/cheat/conf.yml
```
This is where I saw the error. Here are my steps to joy:

```shell
echo export CHEAT_CONFIG_PATH="$HOME/.config/cheat/conf.yml" >> $HOME/.bashrc
echo export CHEAT_PATH="$HOME/AppData/Roaming/cheat/cheatsheets/" >> $HOME/.bashrc
mkdir -p $HOME/AppData/Roaming/cheat/cheatsheets/personal
git clone https://github.com/cheat/cheatsheets.git $HOME/AppData/Roaming/cheat/cheatsheets/community
source $HOME/.bashrc
```
```shell
$ cheat
Usage:
  cheat [options] [<cheatsheet>]

Options:
  --init                  Write a default config file to stdout
  -c --colorize           Colorize output
  -d --directories        List cheatsheet directories
  -e --edit=<cheatsheet>  Edit <cheatsheet>
  -l --list               List cheatsheets
  -p --path=<name>        Return only sheets found on path <name>
  -r --regex              Treat search <phrase> as a regex
  -s --search=<phrase>    Search cheatsheets for <phrase>
  -t --tag=<tag>          Return only sheets matching <tag>
  -T --tags               List all tags in use
  -v --version            Print the version number
  --rm=<cheatsheet>       Remove (delete) <cheatsheet>

Examples:

  To initialize a config file:
    mkdir -p ~/.config/cheat && cheat --init > ~/.config/cheat/conf.yml

  To view the tar cheatsheet:
    cheat tar

  To edit (or create) the foo cheatsheet:
    cheat -e foo

  To edit (or create) the foo/bar cheatsheet on the "work" cheatpath:
    cheat -p work -e foo/bar

  To view all cheatsheet directories:
    cheat -d

  To list all available cheatsheets:
    cheat -l

  To list all cheatsheets whose titles match "apt":
    cheat -l apt

  To list all tags in use:
    cheat -T

  To list available cheatsheets that are tagged as "personal":
    cheat -l -t personal

  To search for "ssh" among all cheatsheets, and colorize matches:
    cheat -c -s ssh

  To search (by regex) for cheatsheets that contain an IP address:
    cheat -c -r -s '(?:[0-9]{1,3}\.){3}[0-9]{1,3}'

  To remove (delete) the foo/bar cheatsheet:
    cheat --rm foo/bar
```

**Comment by chrisallenlane** [comment]
@Benjamin-Connelly, I'm unable to test this, but thank you for the information :slightly_smiling_face:

**Comment by hbrands** [comment]
Hi,
I'm not a GO expert, but could it be a line delimiter issue on Windows (\n vs. \r\n)?
Please have a look at the Prompt function in the installer package:
https://github.com/cheat/cheat/blob/master/internal/installer/prompt.go
reader.ReadString('\n') is used there, but then later
strings.TrimRight(ans, "\n") strips only the \n , but not the '\r' on Windows.
Therefore, false is always returned and no config file created.
strings.TrimRight(ans, "\r\n") should be used instead. 
Just a guess...
What do you think?

**Comment by chrisallenlane** [comment]
@hbrands, while this is a possibility, I don't _think_ this is the issue, simply because (to my knowledge) others are using `cheat` successfully on Windows, and this would, presumably, prevent them from doing so.

That said, to my embarrassment, I am still personally unable to test `cheat` on Windows. I've been relying entirely on the community to do this so far.

I haven't had much time to work on this project recently, because I've been rebuilding most of my computers in my (limited) free time. I think I'm about two weeks away from completing that effort. Fortunately, when that's complete, I'll finally have a Windows PC available for testing, and then will be able to take a deep-dive into this issue.

Once I have that lab machine available, I'll test your hypothesis. While my hunch is that the root cause lies elsewhere, line-endings are worth a look. They are indeed able to cause problems, and I'm not handling them in a particularly enlightened manner so far.

**Comment by hbrands** [comment]
@chrisallenlane Thanks for your answer. Here is a small example that demontrates the problem on my maschine (Windows 10):
```
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {

	// initialize a line reader
	reader := bufio.NewReader(os.Stdin)

	// display the prompt
	fmt.Print(fmt.Sprintf("%s: ", "(y/n) ?"))

	// read the answer
	ans, err := reader.ReadString('\n')
	if err != nil {
		return
	}

	// normalize the answer
	ans1 := strings.ToLower(strings.TrimRight(ans, "\n"))
	fmt.Println("First try:")
	switch ans1 {
	case "y":
		fmt.Println("Good! - Yes!")
	default:
		fmt.Println("Bad!")
	}
	
	// normalize the answer
	ans2 := strings.ToLower(strings.TrimRight(ans, "\r\n"))
	fmt.Println("Second try:")
	switch ans2 {
	case "y":
		fmt.Println("Good! - Yes!")
	default:
		fmt.Println("Bad!")
	}
}
```
The first try always outputs "Bad!" even if I type in "y".
The second try outputs "Good! - Yes!" correctly when I type in "y".

Is there a simple way to build cheat on a windows machine? Then I could try that out.
But I'm pretty sure now that this is the issue.

**Comment by chrisallenlane** [comment]
I just published `cheat` [4.2.4][], which greatly improves Windows support. Please upgrade to that version and try again.

I'll optimistically close this ticket as resolved now, but feel free to re-open it if upgrading doesn't solve your problem.

Thanks.

[4.2.4]: https://github.com/cheat/cheat/releases/tag/4.2.4
